src/app/src/main/kotlin/org/imperial/mrc/hint/FileManager.kt

```diff
@@ -29,46 +32,41 @@ interface FileManager {
 
 @Component
 class LocalFileManager(
-        private val context: WebContext,
-        private val pac4jConfig: Config,
+        private val session: Session,
+        private val stateRepository: StateRepository,
         private val appProperties: AppProperties) : FileManager {
 
     override fun saveFile(file: MultipartFile, type: FileType): String {
-        val id = pac4jConfig.sessionStore.getOrCreateSessionId(context)
-        val fileName = file.originalFilename!!
-        val path = "${appProperties.uploadDirectory}/$id/$type/$fileName"
-        val localFile = File(path)
+        val md = MessageDigest.getInstance("MD5")
+        val bytes = file.inputStream.use {
+            DigestInputStream(it, md).readBytes()
+        }
 
-        if (localFile.parentFile.exists()) {
-            FileUtils.cleanDirectory(localFile.parentFile)
-        } else {
+        val extension = file.originalFilename!!.split(".").last()
+        val hash = "${DatatypeConverter.printHexBinary(md.digest())}.${extension}"
+        val path = "${appProperties.uploadDirectory}/$hash"
+        if (stateRepository.saveNewHash(hash)) {
+            val localFile = File(path)
             FileUtils.forceMkdirParent(localFile)
+            localFile.writeBytes(bytes)
         }
 
-        localFile.writeBytes(file.bytes)
-
+        stateRepository.saveSessionFile(session.getId(), type, hash, file.originalFilename!!)
         return path
     }
 
     override fun getFile(type: FileType): File? {
-
-        val id = pac4jConfig.sessionStore.getOrCreateSessionId(context)
-
-        val pjnzFiles = File("${appProperties.uploadDirectory}/$id/$type")
-                .listFiles()
-
-        return pjnzFiles?.first()
+        val hash = stateRepository.getSessionFileHash(session.getId(), type)
```

Comment:
The db structure allows for multiple so this won't be a problem from that point of view. 

I think the only one that might be multiple is PJNZ? And yes if/when we want to support that then we'd need this method to return a list instead of a singleton. But a fair bit will have to change if/when we support multiple PJNZ but I don't think anything here precludes that.

Not sure what you mean about how to identify them - each file will be uniquely identified by its hash!

Reply:
This is true! I think I was thinking about what parameters you'd pass this method to get one hash out. But I guess you would just get a list of all hashes of that type instead.